aiida_kkr/tools/multi_imps_data_extract.py

```diff
@@ -0,0 +1,453 @@
+                     f'rz: {self.rz},\n'
+                     f'magnetic moment-1: {self.mom1},\n'
+                     f' magnetic moment-2: {self.mom2},\n'
+                     f'total moment: {self.tot_mom}')
+        return print_str 
+
+    def ExtractData(self, node: Node):
+        """Extract data of magnetic impurity couple.
+
+        node: Combine_imps_wc node and data will be extract from this node.
+        """
+        
+        from aiida_kkr.workflows import kkr_imp_sub_wc
+
+        imp_order1 = self.imp_order1
+        imp_order2 = self.imp_order2
+        workflow_info = node.outputs.workflow_info.get_dict()
+        
+        try:
+            imps_info_in_exact_cluster = (workflow_info[
+                                        'imps_info_in_exact_cluster'])
+        except KeyError:
+            imps_info_in_exact_cluster = ExtractImpInfo(node)
+            
+            
+        try:
+            jij_data = node.outputs.JijData.get_array('JijData')
+        except KeyError:
+            raise KeyError(f'The combine_impurity_calc uuid: {node.uuid}'
+                      f' did execute the Jij calculation.')
+            
+        else:
+            jij_info = node.outputs.JijInfo.get_dict()['text']
+
+        impurity_info = (node.get_outgoing(node_class=kkr_imp_sub_wc).all()
+                        [0].node.inputs.impurity_info.get_dict())
+        zimp = impurity_info['Zimp']
+        imp_cls = impurity_info['imp_cls']
+        
+        # To collect the impurity index from cluster 
+        zimp_index_cluster = [ind for ind, cluster_atom_info in 
+                              enumerate(imp_cls) if cluster_atom_info[-2] in 
+                              zimp]
+        self.J_data_uuid = node.outputs.JijData.uuid
+        self.calc_imps = len(zimp_index_cluster)
+        self.imp1 = int(imp_cls[zimp_index_cluster[imp_order1]][-2])
+        self.imp2 = int(imp_cls[zimp_index_cluster[imp_order2]][-2])
+        
+        # For old version of combine_imps_wc 
+        if len(zimp) != len(imps_info_in_exact_cluster['offset_imps']):
+            self.offset1 = 0
+            self.offset2 = imps_info_in_exact_cluster['offset_imps'][
+                                                       imp_order2-1]
+        else:
+            self.offset1 = imps_info_in_exact_cluster['offset_imps'][
+                                                        imp_order1]
+            self.offset2 = imps_info_in_exact_cluster['offset_imps'][
+                                                        imp_order2]
+
+        self.il1 = imps_info_in_exact_cluster['ilayers'][imp_order1]
+        self.il2 = imps_info_in_exact_cluster['ilayers'][imp_order2]
+        
+        jij_info_arr = ([[float(data.strip()) for data in dataline.split(' ')
+                         if data != ''] 
+                         for dataline in jij_info.split('\n')[3:-1]])
+        
+        jij_pairs = [[data[0], data[1]] for data in jij_info_arr]
```

Comment:
why do you take the jij values from the text instead of the ArrayData node directly?